### PR TITLE
ENH: Make errstate a ContextDecorator in Python3

### PR DESCRIPTION
**Pull request by nimish**
Allows errstate to work like a decorator and a context manager in python 3.

More info: https://docs.python.org/3.5/library/contextlib.html#contextlib.ContextDecorator

### CODE REVIEW

File: numpy/core/numeric.py
Comment (mhvk) [inline numpy/core/numeric.py]: I'd have expected a `_errstate_baseclass = ContextDecorator` here?

**Review by mhvk — COMMENTED** [review]
See in-line comment, though I'm guessing this is just a WIP, as it still needs test case and docs.

File: numpy/core/numeric.py
Comment (nimish) [inline numpy/core/numeric.py]: Yep. 🤦‍♂️

**Comment by mhvk** [comment]
I'm sorry, I should have seen this earlier, but as I looked at the existing documentation, I saw that `errstate` *is already* a context manager! Indeed, it has its own `__enter__` and `__exit__` methods. So, I think this change is not necessary.... Am I missing something?

**Comment by nimish** [comment]
@mvhk While `errstate` is a context manager, it can't be used as a decorator to wrap entire functions --  deriving from the `contextlib.ContextDecorator` base class in contextlib solves this mildly irritating issue.

Alternatively, using the `contextlib.contextmanager` decorator and rewriting errstate to be a generator with state would work as well, without needing the version check. In Python 3, it would magically inherit from `contextlib.ContextDecorator`

File: numpy/core/numeric.py
Comment (eric-wieser) [inline numpy/core/numeric.py]: I'd prefer this in the `else` clause

**Comment by eric-wieser** [comment]
This seems pretty reasonable. I'd almost be tempted just to provided a shim in `compat.py` if we're going to do this though, since it's only [six lines of code](https://github.com/python/cpython/blob/3.6/Lib/contextlib.py#L48-L53)

**Comment by eric-wieser** [comment]
> Alternatively, using the contextlib.contextmanager decorator and rewriting errstate to be a generator with state would work as well, without needing the version check. In Python 3, it would magically inherit from contextlib.ContextDecorator

I like this option more, I think

**Comment by nimish** [comment]
@eric-wieser personally I think we're veering into bike-shedding territory here. Rewriting it as a generator is a pretty invasive change and generators are poorly understood by most people including me (i don't think there's a need for that try/finally since yield won't fail for good cases, and there's no `self` in a standalone generator). The decorator that turns it into a context manager also does some tricky code and it just seems a bit too clever for me.

The current PR is the minimal change that improves the current, working code with the convenience of being a decorator. I suggest we merge it and then consider a rewrite as a generator only if it's worth it for maintainability.

**Comment by eric-wieser** [comment]
> The decorator that turns it into a context manager also does some tricky code and it just seems a bit too clever for me.

That's not relevant though - the decision was made to include that in the standard library, so we needn't concern ourselves about the complexity of its implementation

> generators are poorly understood by most people including me

No understanding of generators is needed here - the documentation of `contextlib.contextmanager` explains exactly what `yield` does in this context.

> i don't think there's a need for that try/finally since yield won't fail for good cases

You're wrong. From the documentation of `contextlib.contextmanager`:
>  If an unhandled exception occurs in the block, it is reraised inside the generator at the point where the yield occurred. Thus, you can use a try...except...finally statement to trap the error (if any), or ensure that some cleanup takes place.

**Comment by nimish** [comment]
QED!

You in fact *do* need to understand the complex implementation of `contextlib.contextmanager` since that's not the normal behavior of yield, and `GeneratorContextManager` is effectively doing some tricky code under the hood (instead of, say, passing it as the argument as a coroutine).  

Just take a look at the implementation [here](https://github.com/python/cpython/blob/master/Lib/contextlib.py#L56) -- there's more comments than actual code because it's difficult to understand and debug, AND there's a [bug](http://bugs.python.org/issue19404) on the docs. It's an inferior implementation compared to our current, *working*, one.

If we were writing `errstate` from scratch I think I would be more receptive to going the generator route, but good engineering practice indicates that we shouldn't change things for the sake of it.

**Comment by eric-wieser** [comment]
> You in fact do need to understand the complex implementation of contextlib.contextmanager

Let's not confuse implementation and API. You need to understand the contents of the docstring. You do not need to understand the implementation to use it.

> AND there's a bug on the docs

This bug only affects `help(np.errstate(...))`, which would be a very unusual way to ask for help (compared to `help(np.errstate)`). Also, we already fall afoul of this bug in numpy with `ufunc.__doc__`. which is why `np.info` exists

**Comment by nimish** [comment]
OK, well, then the API is bizarre and too-clever-by-half -- I guarantee some idiot like me will complain that there's an "unneeded" try/finally but due to the peculiar and non-standard generator you do need it. I'm not arguing that point

I *am* arguing that there's not a compelling reason to rewrite working code unless necessary or there's a meaningful improvement, and this exchange alone is a good argument against the maintainability of the generator implementation. The class based approach has the advantage of being easy to understand, debug and doesn't have the doc-bug at all, regardless of how unlikely it is to be triggered in practice.

I regret bringing it up, frankly.

If you feel strongly about it I will accept a PR on my PR 😄 

Regardless: what is blocking this PR from being merged, aside from this?

**Comment by eric-wieser** [comment]
> what is blocking this PR from being merged, aside from this?

This isn't blocking merging - but I've an nitpick in an inline comment above.

**Comment by eric-wieser** [comment]
Also needs a release note

File: numpy/core/numeric.py
Comment (eric-wieser) [inline numpy/core/numeric.py]: Needs `..versionchanged: 1.14` before, or whatever the actual syntax is. Can we mention `ContextDecorator` here?

Comment (eric-wieser) [inline numpy/core/numeric.py]: Or better yet, do an intersphinx link, which I forget how to do

**Comment by nimish** [comment]
@eric-wieser Where/how do I add a release note for 1.14.0 ?

File: numpy/core/numeric.py
Comment (eric-wieser) [inline numpy/core/numeric.py]: Suspect you might need a blank line here, or that the following section needn't be indented - but I'm not at all sure

**Comment by eric-wieser** [comment]
`doc\release\1.14.0-notes.rst`, under Improvements, I think

File: numpy/core/numeric.py
Comment (nimish) [inline numpy/core/numeric.py]: Looks ok according to http://www.sphinx-doc.org/en/stable/markup/para.html ?

File: numpy/core/numeric.py
Comment (eric-wieser) [inline numpy/core/numeric.py]: To confirm my hunch, it doesn't appear that we indent this description in any of the other uses in numpy - so perhaps keep the longer multi-paragraph description?

Comment (nimish) [inline numpy/core/numeric.py]: Really? I searched for `versionchanged` and all 3 uses are indented. I'll restore the extended description since https://github.com/numpy/numpy/blob/0767d38c8502aab52424ff9907ac1b63dc138597/numpy/ma/core.py#L5266 has something similar

Comment (eric-wieser) [inline numpy/core/numeric.py]: Turns out I was looking at `versionadded`, which is not used like that.

File: doc/release/1.14.0-notes.rst
Comment (eric-wieser) [inline doc/release/1.14.0-notes.rst]: I think this needs 3 more `-` to render correctly

**Review by eric-wieser — APPROVED** [review]
Generally LGTM

File: doc/release/1.14.0-notes.rst
Comment (nimish) [inline doc/release/1.14.0-notes.rst]: Fixed

**Comment by mhvk** [comment]
I'm happy with this too. I guess the only thing to do is to squash all the commits together.

**Comment by nimish** [comment]
@mvhk Just squash merge it and I think we're good to go: https://help.github.com/articles/about-pull-request-merges/

**Comment by nimish** [comment]
@mhvk @eric-wieser Anything left to do on this? I squashed the commits.

**Comment by seberg** [comment]
I know it might be annoying code wise, but it would not take much to make this python 2.7 compatible, no? It would seem preferable to me.

**Comment by eric-wieser** [comment]
Unfortunately this still isn't py2 compatible, but it does at least not special case py3, so I prefer it to the previous patch.

I don't think it's within scope of numpy to backport contextlib

**Comment by seberg** [comment]
Yeah, but it is not a big deal to just add a `__call__` to the class so that it can be used as a decorator, we have a few things like that already.

**Comment by nimish** [comment]
@seberg @eric-wieser I really don't care about py2 support since it's going to be EoL reasonably soon and I have no need for it. 

Can someone make a decision on what's required to merge this? 

I can reinstate the py3 specific doc changes and test guard.

File: numpy/core/numeric.py
Comment (eric-wieser) [inline numpy/core/numeric.py]: I think if this is a py3 only feature, we shouldn't emphasize this - simply mentioning that it uses `contextlib.contextmanager`, and so inherits any new features that entails is probably the way to go. Similar for the release notes.

**Comment by charris** [comment]
Conflict easy to fix. Python 2 failures need to be dealt with.

**Comment by nimish** [comment]
Finally had some more time to work on this. Options:

1. Leave it py3 only 
    - easy, no changes for py2.7 folks
    - py 2.7 is close to EoL so spending time and effort doing this right is a waste

2. Reimplement stuff to make it py2.7 compatible 
    - irritating backwards compat
    - If someone can write the necessary code, I am ok with merging it in to this change

Personally, I would merge the py3 code and then we can work in py2.7 code if someone wishes to backport contextlib. At the very least, py3 people should be able to use this feature as soon as possible.

File: doc/release/1.14.0-notes.rst
Comment (eric-wieser) [inline doc/release/1.14.0-notes.rst]: Merge conflict here. Needs to move to the 1.15 release notes anyway - sorry for not being faster on this

File: numpy/core/numeric.py
Comment (eric-wieser) [reply numpy/core/numeric.py]: Version needs to change to 1.15 here too

**Comment by charris** [comment]
Pushing off. @nimish If you don't want to support 2.7 this will need to wait until next year.

**Comment by charris** [comment]
Needs a rebase and the release note needs to go elsewhere. Should we just wait until we drop Python 2.7? We have a label for that.

**Comment by mattip** [comment]
@nimish Can you update this?

- the release note should be on `doc/release/1.17.0-notes.rst`
- rebase, fix conflicts and squash
- since HEAD is currently broken you should rebase to the last good commit which is e6970fd7

**Comment by nimish** [comment]
> @nimish Can you update this?
> 
> * the release note should be on `doc/release/1.17.0-notes.rst`
> * rebase, fix conflicts and squash
> * since HEAD is currently broken you should rebase to the last good commit which is [e6970fd](https://github.com/numpy/numpy/commit/e6970fd7e4c06c8fea19ed4c927b2918dccd6801)

@mattip yes I will have some time this week. Will do.

**Comment by mattip** [comment]
@nimish ping

**Comment by nimish** [comment]
Sorry -- was ill! Is the latest head still broken?

**Comment by nimish** [comment]
@mattip @charris @mhvk  -- any further fixes? it looks like the only conflict is not a real one, and should be done by the maintainers on merge

File: numpy/core/numeric.py
Comment (mhvk) [inline numpy/core/numeric.py]: Right now, `_errstate_baseclass` exists only because of python2; should this be the first file where we just remove the python2 work-arounds?

File: numpy/core/tests/test_errstate.py
Comment (mhvk) [inline numpy/core/tests/test_errstate.py]: If this addition still needed?

Comment (nimish) [reply numpy/core/tests/test_errstate.py]: 🤷‍♂️ https://github.com/numpy/numpy/blob/b047f387d2424bd2134db6005190311b15a84ac2/doc/TESTS.rst.txt

Says it's OK, but the notes for 1.14.0 says it'll break. Might wanna sort that out.

File: numpy/core/numeric.py
Comment (nimish) [reply numpy/core/numeric.py]: Are you suggesting I do this or not?

Comment (mhvk) [reply numpy/core/numeric.py]: Hmm, yes, I am, but am not sure @charris would agree this is a good way to proceed.

Certainly, it is fine *not* to do this right now.

File: doc/release/1.17.0-notes.rst
Comment (eric-wieser) [inline doc/release/1.17.0-notes.rst]: Missing word at the end here

File: numpy/core/numeric.py
Comment (charris) [reply numpy/core/numeric.py]: We will not immediately remove workarounds from existing code, but new code doesn't need them unless it is planned for a backport.

File: numpy/core/tests/test_errstate.py
Comment (charris) [reply numpy/core/tests/test_errstate.py]: Not wanted, and wouldn't work anyway without importing `run_module_suite`, which only works for Nose.

Comment (charris) [reply numpy/core/tests/test_errstate.py]: The documentation needs a fix, @mattip ping.

Comment (nimish) [reply numpy/core/tests/test_errstate.py]: removed.

File: doc/release/1.17.0-notes.rst
Comment (nimish) [reply doc/release/1.17.0-notes.rst]: fixed.

File: numpy/core/numeric.py
Comment (nimish) [reply numpy/core/numeric.py]: @charris so should this workaround be removed?

Comment (mhvk) [reply numpy/core/numeric.py]: @charris - here, there is a specific work-around, as the whole PR doesn't work on python3, but removing it requires also removing other python2 work-arounds. My sense is that we might as well get started (if @nimish is willing), but we can also take the easy route and just go with what is here.

Comment (charris) [reply numpy/core/numeric.py]: What is wrong with just using `ContextDecorator`

> as the whole PR doesn't work on python3

I thought this work was all Python3 only?

Comment (mhvk) [reply numpy/core/numeric.py]: Indeed, it is, but right now up at the top of the file there are separate paths for imports for python3 and python2 (with `_errstate_baseclass` set to `object` in python2, since the `errstate` class itself works fine on python2. So, the specific question is indeed: should we just use `ContextDecorator` here, and, since that will break usage on python2, also remove the python2-specific path at the top.

File: numpy/core/tests/test_errstate.py
Comment (mattip) [reply numpy/core/tests/test_errstate.py]: @charris #12787

File: numpy/core/numeric.py
Comment (charris) [reply numpy/core/numeric.py]: I'd just import `ContextDecorator` outside of the if statement and leave it alone otherwise. We are keeping the Python 2 workarounds to simplify backports, but I don't think we need to keep them current :)

**Comment by eric-wieser** [comment]
Put another way, there would be little reason to backport this. 1.16 _is_ a LTS release, but realistically Python 3 users are unlikely to stick with it once 1.17 is out

**Comment by nimish** [comment]
OK, I removed the py2 fallback. It just imports and uses `ContextDecorator` directly. @mattip @charris  @eric-wieser thoughts?

**Comment by charris** [comment]
I went ahead and fixed up the release notes conflict.

File: numpy/core/numeric.py
Comment (nimish) [reply numpy/core/numeric.py]: That's done.

**Comment by mattip** [comment]
Thanks @nimish

**Pull request merged**
